When adopting a thread, spin until GC isn't running.

maleadt opened the pull request:
Fixes https://github.com/JuliaLang/julia/issues/49928, alternative to https://github.com/JuliaLang/julia/pull/49930.

maleadt committed:
When adopting a thread, spin until GC isn't running.

maleadt committed:
Remove NOTSAFEPOINT annotation from jl_adopt_thread.

maleadt commented:
I'm not sure how to fix the analyzer error; `jl_adopt_thread` is currently annotated `JL_NOTSAFEPOINT_LEAVE` (what does the `LEAVE` mean?), but many of the functions it uses call into the GC. I guess the analyzer wasn't visiting this function before?

```
julia/src/threading.c:424:21: error: Calling potential safepoint as SimpleFunctionCall from function annotated JL_NOTSAFEPOINT [julia.GCChecker]
    jl_task_t *ct = jl_init_root_task(ptls, stack_lo, stack_hi);
                    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
julia/src/julia_internal.h:863:12: note: Tried to call method defined here
jl_task_t *jl_init_root_task(jl_ptls_t ptls, void *stack_lo, void *stack_hi);
           ^~~~~~~~~~~~~~~~~
```

We can't annotate `jl_init_root_task` `JL_NOTSAFEPOINT` as it calls `jl_gc_alloc`.

vchuravy commented:
This only fixes part of the issue. You still need to fix the issue where we get past this check and then GC starts and we hit a safepoint without being able to execute that safepoint on this thread yet.

maleadt committed:
Disable GC during thread adoption.

Otherwise GC might run after the foreign thread exits the spin loop,
but before it is initialized and ready to handle GC safepoints.

vchuravy commented on src/threading.c:
```diff
@@ -408,16 +408,24 @@ jl_ptls_t jl_init_threadtls(int16_t tid)
+    while (jl_atomic_load_relaxed(&jl_gc_running) || jl_atomic_load_acquire(&jl_gc_running)) {
+        jl_cpu_pause();
+    }
+    jl_atomic_fetch_add(&jl_gc_disable_counter, 1);
```
Can you change the order here?

1. Request GC is off
2. Wait for any GC to finish

There is still a window I am worried about. In `jl_gc_collect` we:

1. Check if GC is allowed to run
2. Win the race to become GC thread and mark gc_running

So there is a possible interleaving.

Where 
1. Check GC is allowed to run
2. Foreign Thread requests GC off
3. Foreign Thread waits for GC to finish
4. Win the race to become GC thread and mark gc_running

vchuravy replied:
Thinking about it some more we will need to check disable_counter after winning the race

maleadt committed:
Check jl_gc_disable_counter again before arming the safepoint to prevent a data race.

vchuravy commented on src/threading.c:
```diff
@@ -408,16 +408,24 @@ jl_ptls_t jl_init_threadtls(int16_t tid)
     jl_init_stack_limits(0, &stack_lo, &stack_hi);
 
-    (void)jl_gc_unsafe_enter(ptls);
     // warning: this changes `jl_current_task`, so be careful not to call that from this function
```
```suggestion
    // warning: this changes `jl_current_task`, so be careful not to call that from this function
    // `jl_init_root_task` assumes that GC is disabled.
```

vchuravy commented on src/safepoint.c:
```diff
@@ -124,6 +124,13 @@ int jl_safepoint_start_gc(void)
+    // Foreign thread adoption disables the GC and waits for it to finish, however, that may
+    // introduce a race between it and this thread checking if the GC is enabled and only
+    // then setting jl_gc_running. To avoid that, check again now that we won that race.
+    if (jl_atomic_load_relaxed(&jl_gc_disable_counter)) {
```
Uhm do we have to make this `acquire`

maleadt replied:
I'm not sure. Probably? @vtjnash
If so, I guess the code I copied this from in `jl_gc_collect` should be updated too?

vchuravy replied:
I think the issue is that if the check is after the safepoint we have a memory-fence/happens-before-relationship established.

Here we don't have a happens-before relationship with that particular thread since it will not participate in the stop-the world scenario (and this is before the safepoint). 

I tend to go back to https://marabos.nl/atomics/memory-ordering.html#happens-before to refresh my memory.

vchuravy commented on src/threading.c:
```diff
@@ -408,16 +408,24 @@ jl_ptls_t jl_init_threadtls(int16_t tid)
+    // jl_init_threadtls puts us in a GC unsafe region, so ensure GC isn't running.
+    // we can't use a normal safepoint because we don't have signal handlers yet.
+    // we also can't use jl_safepoint_wait_gc because that assumes we're in a task.
+    jl_atomic_fetch_add(&jl_gc_disable_counter, 1);
```
and this release?

maleadt replied:
`jl_atomic_fetch_add` is implemented using `std::atomic_fetch_add`, and AFAICT with no order specified that uses seq_cst, so should be fine. And our `_explicit` versions take a C++ ordering, so I can't use those from `threading.c`.

vchuravy commented:
Besides the discussion around memory model this looks.good to me.

I will leave final review to @vtjnash and @JeffBezanson

gbaraldi commented:
What happens if a thread calls `GC.enable()` and tries to enter GC in the meanwhile?

gbaraldi commented:
Oh, I see it now we have a per thread check.

maleadt committed:
Update src/threading.c

Co-authored-by: Valentin Churavy <vchuravy@users.noreply.github.com>

maleadt committed:
Update src/threading.c

Co-authored-by: Valentin Churavy <vchuravy@users.noreply.github.com>

maleadt committed:
Change loads of jl_gc_disable_counter to use acquire semantics.

Stores already use seq_cst.

maleadt commented:
Couple of CI hangs, with one printing the following backtrace:

```
KILLING BY THREAD TEST WATCHDOG
======================================================================================
Information request received. A stacktrace will print followed by a 1.0 second profile
======================================================================================
cmd: /cache/build/default-armageddon-1/julialang/julia-master/julia-c2ef05e211/bin/julia 9614 running 2 of 5
signal (10): User defined signal 1
pthread_cond_wait at /lib/aarch64-linux-gnu/libpthread.so.0 (unknown line)
uv_cond_wait at /workspace/srcdir/libuv/src/unix/thread.c:883
jl_safepoint_wait_gc at /cache/build/default-armageddon-7/julialang/julia-master/src/safepoint.c:172
jl_safepoint_start_gc at /cache/build/default-armageddon-7/julialang/julia-master/src/safepoint.c:124
ijl_gc_collect at /cache/build/default-armageddon-7/julialang/julia-master/src/gc.c:3513
gc at ./gcutils.jl:129 [inlined]
macro expansion at /cache/build/default-armageddon-1/julialang/julia-master/julia-c2ef05e211/share/julia/test/threads_exec.jl:209 [inlined]
#251#threadsfor_fun#21 at ./threadingconstructs.jl:184
#251#threadsfor_fun at ./threadingconstructs.jl:151 [inlined]
#1 at ./threadingconstructs.jl:129
unknown function (ip: 0xffffa1e52f4b)
```

So that seems related.

maleadt committed:
Clear jl_gc_running when backing out.

maleadt committed:
Rephrase comments.

vtjnash commented on src/safepoint.c:
```diff
@@ -124,6 +124,14 @@ int jl_safepoint_start_gc(void)
+    // introduce a race between it and this thread checking if the GC is enabled and only
+    // then setting jl_gc_running. To avoid that, check again now that we won that race.
+    if (jl_atomic_load_acquire(&jl_gc_disable_counter)) {
+        jl_atomic_store_release(&jl_gc_running, 0);
```
This does not seem logical to be a release, since there are no other stores in this function to act upon

vtjnash replied:
I looks like you intended the much more expensive seq-cst semantics so that it establishes an atomic ordering with the prior load?

maleadt replied:
I'm not sure. The load during `jl_adopt_thread` is an atomic acquire, which I intended to match with this release store. I'm not sure why it would need to be matched up with stores in this very function?
